lib/stages/mergeBeamBuffer.cpp

```diff
@@ -0,0 +1,134 @@
+            out_metadata->sub_frame_data_size = in_buf->frame_size;
+        }
+
+        // Get the start position of a sub_out_frame's meatadata
+        // Make sure out_frame is pointed to the right place.
+        if (out_frame == nullptr)
+            break;
+        uint32_t out_frame_metadata_pos = sub_frame_size * sub_frame_index;
+        // Get the memory block for the metadata of sub_out_frame.
+        FreqIDBeamMetadata* sub_frame_metadata =
+            (FreqIDBeamMetadata*)&out_frame[out_frame_metadata_pos];
+        // Copy the in_frame metadata the the out frame's subframe metadata
+        // Using memory copy here will give a warning.
+        sub_frame_metadata->fpga_seq_start = in_metadata->fpga_seq_start;
+        sub_frame_metadata->ctime = in_metadata->ctime;
+        sub_frame_metadata->stream_id = in_metadata->stream_id;
+        sub_frame_metadata->dataset_id = in_metadata->dataset_id;
+        sub_frame_metadata->beam_number = in_metadata->beam_number;
+        sub_frame_metadata->ra = in_metadata->ra;
+        sub_frame_metadata->dec = in_metadata->dec;
+        sub_frame_metadata->scaling = in_metadata->scaling;
+
+        // Add the frequency id into the metadata, for chime one stream has on frequency.
+        // TODO Have the output buffer handle the stream with mulitple frequency
+        const uint32_t num_freq_per_stream = Telescope::instance().num_freq_per_stream();
+
+        for (uint32_t f = 0; f < num_freq_per_stream; ++f) {
+            // TODO this does not handling the stream with mulitple frequencies. since at right now,
+            // num_freq_per_stream is 1
+            sub_frame_metadata->frequency_bin =
+                Telescope::instance().to_freq_id(in_metadata->stream_id, f);
+        }
+
+        // Get the start position of a sub_out_frame's data
+        uint32_t out_frame_data_pos = out_frame_metadata_pos + sizeof(FreqIDBeamMetadata);
+        // Get the memory block for the data of of a subframe.
+        uint8_t* sub_frame_data = &out_frame[out_frame_data_pos];
+        // copy data from in frame to sub_out_frame
+        memcpy(sub_frame_data, in_frame, in_buf->frame_size);
+
+        // Debug code. Check if the subframe matedata is the same.
+        FreqIDBeamMetadata* debud_frame_metadata =
+            (FreqIDBeamMetadata*)&out_frame[sub_frame_index * sub_frame_size];
+        DEBUG2("Sub frame Beam RA: {:f}, Dec: {:f}, scaling: {:d}, beam_num: {:d}, freq_id {:d}\n",
+               debud_frame_metadata->ra, debud_frame_metadata->dec, debud_frame_metadata->scaling,
+               debud_frame_metadata->beam_number, debud_frame_metadata->frequency_bin);
+
+        mark_frame_empty(in_buf, unique_name.c_str(), in_buffer_ID);
+        in_buffer_ID++;
+        in_frame_counter++;
+
+        if (sub_frame_index == _sub_frames_per_merged_frame - 1) {
+            mark_frame_full(out_buf, unique_name.c_str(), out_buffer_ID);
+            out_buffer_ID++;
```

Comment:
Minor comment: you might want to set `out_frame == nullptr;` here, just in case the logic above didn't work for some reason it would seq fault instead of writing to an old pointer.